### DISCUSSION TITLE
Is it possible to access variables defined on the jupyterlite site through the console?

### DISCUSSION THREAD

**Discussion by savakarrohan** [jupyterlite/jupyterlite#946] [Q&A]
![image](https://user-images.githubusercontent.com/50200523/211814934-1de8ca58-046f-4193-aee1-c7de9c91dfab.png)

I want to access the variable `a` from the console on a browser tab. 
Do you know how I can go about this?
I am attempting to build an extension to Jupyter notebooks that accesses these variables from python. 
The extension will be written in typescript/javascript according to the various examples shown in the tutorials.

I attempted to run the following code through the console 
just going through the pyodide documentation [here](https://pyodide.org/en/stable/usage/quickstart.html#accessing-python-scope-from-javascript)
` pyodide.globals.get('a').toJs();`
gave me an error which says pyodide is not loaded. which kinda makes sense. but yeah any headway will help.

**Comment by jtpio** [comment]
The Pyodide kernel currently runs in a web worker, which means that `pyodide` is indeed not available at the top level in the dev tools console.

@savakarrohan out of curiosity, what would be the purpose of retrieving the value of the Python variable in the dev tools console?

**Reply by savakarrohan** [reply]
So I was hoping to write an extension that will be able to convert a "flowchart" into the code. 
I just like python, and the jupyterlabs UI is AMAZING. 

I feel like a majority of people don't attempt to codify their decisions. The ability to run python without any installation/env/packages is definitely fun and simplifies a ton of items. 

I imagine being able to run flowcharts online.

I have been lurking around the jupyter group for 4-6 months now. I understand that extensions are written in typescript. Yes, I can do a pure react tool but my liking of python and the various libraries in this "high-level language" just makes me want to do it using python.

But I imagine also being able to use open-source famous libraries such as numpy, scipy and pandas functions. Not sure how yet but maybe each function will be like a rectangle box doing a process and having fixed outputs ¯\_ (ツ)_/¯ with lines displaying the output.

Thank you so much for asking, its helped me make my idea a tad bit more focused now.

![image](https://user-images.githubusercontent.com/50200523/212174489-20e92b92-88b9-44fa-89e2-5a8652172835.png)

**Reply by savakarrohan** [reply]
Oh, by the way, I just thought if I was able to access the variable on the console, I would be able to access it through the extension code in typescript ...

**Reply by jtpio** [reply]
> Oh, by the way, I just thought if I was able to access the variable on the console, I would be able to access it through the extension code in typescript ...

Yes in that case it would be true, since the extension code (TypeScript) would be running in the main window thread.

It sounds like this could be done using the Jupyter messaging protocol and sending messages to the kernel to evaluate variables: https://jupyter-client.readthedocs.io/en/stable/messaging.html

A good candidate would have been the `inspectVariables` debug message, although it's not currently not supported in JupyterLite.

This example might be relevant: https://github.com/jupyterlab/extension-examples/tree/master/kernel-messaging

**Reply by savakarrohan** [reply]
This seems like what I would like, I shall look into it. Thanks.

**Comment by qqdaiyu55** [comment]
You can save variables to indexDB in notebook: [code snippets](https://github.com/jupyterlite/jupyterlite/discussions/91?sort=new),
and access the indexDB data in extension codes. To build data transferring between main thread & service worker, indexDB could be a quick option.

**Reply by savakarrohan** [reply]
I should look into this, Thank you!
